crates/core/src/metadata_store/providers/objstore/optimistic_store.rs

```diff
@@ -0,0 +1,254 @@
+
+pub(crate) struct OptimisticLockingMetadataStoreBuilder {
+    pub(crate) version_repository: Arc<dyn VersionRepository>,
+    pub(crate) _configuration: MetadataStoreClient,
+}
+
+impl OptimisticLockingMetadataStoreBuilder {
+    pub(crate) async fn build(&self) -> anyhow::Result<OptimisticLockingMetadataStore> {
+        let refresh_duration = Duration::from_secs(5);
+        Ok(OptimisticLockingMetadataStore::new(
+            refresh_duration,
+            self.version_repository.clone(),
+        ))
+    }
+}
+
+struct CachedStore {
+    store: ImmutableStore,
+    created_at: Instant,
+}
+
+impl CachedStore {
+    fn new(store: ImmutableStore) -> Self {
+        Self {
+            store,
+            created_at: Instant::now(),
+        }
+    }
+
+    fn is_stall(&self, duration: Duration) -> bool {
+        self.created_at.elapsed() > duration
+    }
+}
+
+impl Default for CachedStore {
+    fn default() -> Self {
+        Self {
+            store: Default::default(),
+            created_at: Instant::now(),
+        }
+    }
+}
+
+pub struct OptimisticLockingMetadataStore {
+    version_repository: Arc<dyn VersionRepository>,
+    latest_store_cache: ArcSwap<Option<CachedStore>>,
+    refresh_interval: Duration,
+}
+
+impl OptimisticLockingMetadataStore {
+    fn new(refresh_interval: Duration, version_repository: Arc<dyn VersionRepository>) -> Self {
+        Self {
+            version_repository,
+            refresh_interval,
+            latest_store_cache: Default::default(),
+        }
+    }
+
+    async fn maybe_update_cached_store(&self, force_update: bool) -> anyhow::Result<()> {
+        let needs_refresh = force_update
+            || self
+                .latest_store_cache
+                .load()
+                .as_ref()
+                .as_ref()
+                .map(|store| store.is_stall(self.refresh_interval))
+                .unwrap_or(true);
+
+        if !needs_refresh {
+            return Ok(());
+        }
+
+        let maybe_global_latest_version = self
+            .version_repository
+            .get_latest_version()
+            .await
+            .map_err(|e| anyhow!(e))?;
+
+        let maybe_cached_version = self
+            .latest_store_cache
+            .load()
+            .as_ref()
+            .as_ref()
+            .map(|cache| cache.store.current_version());
+
+        let Some(global_latest_version) = maybe_global_latest_version else {
+            // nothing to do, currently there is no known (or visible) latest version.
+            // we will optimistically create an empty store, assuming that there is no
+            // previous metadata stored.
+            if maybe_cached_version.is_none() {
+                self.latest_store_cache
+                    .store(Arc::new(Some(CachedStore::default())));
+            }
+            return Ok(());
+        };
+
+        if maybe_cached_version.is_none() || maybe_cached_version.unwrap() < global_latest_version {
```

Comment:
Not sure whether it's easier to read: `if !maybe_cached_version.is_some_and(|cached_version| cached_version >= global_latest_version)`

Reply:
I prefer the `!is_some_and(...)` version personally.